ui/de-lib/src/main/java/org/iplantc/de/admin/desktop/client/toolAdmin/service/impl/ToolAdminServiceFacadeImpl.java

```diff
@@ -0,0 +1,111 @@
+    @Inject private DiscEnvApiService deService;
+
+    @Override
+    public void getTools(String searchTerm, AsyncCallback<List<Tool>> callback) {
+        String address = TOOLS + "?search=" + URL.encodeQueryString(searchTerm);
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(GET, address);
+        deService.getServiceData(wrapper, new ToolListCallbackConverter(callback, factory));
+    }
+
+    @Override
+    public void getToolDetails(String toolId, AsyncCallback<Tool> callback) {
+        String address = TOOLS + "/" + toolId;
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(GET, address);
+        deService.getServiceData(wrapper, new ToolCallbackConverter(callback, factory));
+    }
+
+    @Override
+    public void addTool(ToolList toolList, AsyncCallback<Void> callback) {
+        String address = TOOLS_ADMIN;
+
+        final Splittable encode = AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(toolList));
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(POST, address, encode.getPayload());
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    @Override
+    public void updateTool(Tool tool, boolean overwrite, AsyncCallback<Void> callback) {
+
+        String toolId = tool.getId();
+        String address;
+        if (overwrite) {
+            address = TOOLS_ADMIN + "/" + toolId + "?overwrite-public=true";
+        } else {
+            address = TOOLS_ADMIN + "/" + toolId;
+        }
+
+        final Splittable encode = AutoBeanCodex.encode(AutoBeanUtils.getAutoBean(tool));
+        String payload = removeEmptyValues(encode);
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(PATCH, address, payload);
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    @Override
+    public void deleteTool(String toolId, AsyncCallback<Void> callback) {
+        String address = TOOLS_ADMIN + "/" + toolId;
+
+        ServiceCallWrapper wrapper = new ServiceCallWrapper(DELETE, address);
+        deService.getServiceData(wrapper, new StringToVoidCallbackConverter(callback));
+    }
+
+    private String removeEmptyValues(Splittable payload) {
+        removeEmptyValuesHelper(payload);
+        return payload.getPayload();
+    }
+
+    private void removeEmptyValuesHelper(Splittable obj) {
+        List<String> propertyKeys = obj.getPropertyKeys();
+
+        for (String key : propertyKeys) {
+            Splittable value = obj.get(key);
+            String temp = value.getPayload().replaceAll("[\\[\\]\\{\\}]", "");
+            if ((temp.equals("") || temp.equals("\"\"")) && !key.matches("(input|output)_files")) {
+                StringQuoter.nullValue().assign(obj, key);
+            } else {
+                if (value.isIndexed() || value.isKeyed()) {
+                    removeEmptyValuesHelper(value);
+                }
+            }
+        }
+    }
+}
```

Comment:
Does this also apply to the outermost fields in Tool?  It seems if I set a tool's `attribution`, for example, to some value, then update it to null, it gets omitted from the outgoing JSON body.  Then when I pull up that same tool, the old value for `attribution` is still there.